core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -63,6 +65,8 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
         doWork()
       }
     } catch{
+      case fatalExitError: FatalExitError =>
+        fatalExitError.shutdownSystem()
```

Comment:
The solution from KAFKA-3924 slightly different though uses the same idea of separate thread.
We cannot run shutdown hook in scheduler thread pool because KafkaServer.shutdown() waits until it finishes. Any thread in scheduler thread pool is known by Kafka.